samza-core/src/main/java/org/apache/samza/execution/JobGraphJsonGenerator.java

```diff
@@ -0,0 +1,244 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.execution;
+
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.samza.operators.MessageStream;
+import org.apache.samza.operators.MessageStreamImpl;
+import org.apache.samza.operators.spec.OperatorSpec;
+import org.apache.samza.operators.spec.PartialJoinOperatorSpec;
+import org.apache.samza.operators.spec.SinkOperatorSpec;
+import org.codehaus.jackson.annotate.JsonProperty;
+import org.codehaus.jackson.map.ObjectMapper;
+
+/**
+ * This class generates the JSON representation of the {@link JobGraph}.
+ */
+public class JobGraphJsonGenerator {
+
+  /**
+   * This class provides the necessary connection of operators for traversal.
+   */
+  static abstract class Traversable {
+    @JsonProperty("NextOperatorIds")
+    Set<Integer>  nextOperatorIds = new HashSet<>();
+  }
+
+  static final class OperatorJson extends Traversable {
+    @JsonProperty("OpCode")
+    String opCode;
+    @JsonProperty("OpId")
+    int opId;
+    @JsonProperty("OutputStreamId")
+    String outputStreamId;
+    @JsonProperty("PairedOpId")
+    int pairedOpId = -1;  //for join operator, we will have a pair nodes for two partial joins
+  }
+
+  static final class StreamSpecJson {
+    @JsonProperty("Id")
+    String id;
+    @JsonProperty("SystemName")
+    String systemName;
+    @JsonProperty("PhysicalName")
+    String physicalName;
+    @JsonProperty("PartitionCount")
+    int partitionCount;
+  }
+
+  static final class StreamEdgeJson {
+    @JsonProperty("StreamSpec")
```

Comment:
Let's add source and destination jobIds as well to represent the connectivity information.